File: 03_Duplicate_Numbers_in_an_array/Golang/main.go
package main

func findRepeatNumber(nums []int) int {
	i := 0
	for {
		if i >= len(nums) {
			break
		}

		if nums[i] == i {
			i++
			continue
		}

		if nums[nums[i]] == nums[i] {
			return nums[i]
		}

		tem := nums[nums[i]]
		nums[nums[i]] = nums[i]
		nums[i] = tem

		// nums[i], nums[nums[i]] = nums[nums[i]], nums[i]
	}

	return -1
}

func main() {
	nums := []int{2, 3, 1, 0, 2, 5, 3}
	findRepeatNumber(nums)
}

File: 03_Duplicate_Numbers_in_an_array/python/main.py
class Solution:
    def findRepeatNumber(self, nums: List[int]) -> int:
        i = 0
        while True:
            if nums[i] == i:
                i += 1
                continue
            
            if nums[nums[i]] == nums[i]:
                return nums[i]

            self.swap(nums, i, nums[i])

            
    def swap(self, nums, lindex, rindex):
        a = nums[lindex]
        nums[lindex] = nums[rindex]
        nums[rindex] = a 

